src/main/datalog_console/components/tree_table.cljs

```diff
@@ -1,57 +1,72 @@
   (let [open? (r/atom false)]
     (fn []
       [:<>
        [:tr {:class "odd:bg-gray-100"}
         (doall
          (for [[i col] (map-indexed vector row)]
-           ^{:key (str {:level level :col col})}
-           [:<>
-            (if (= 0 i)
-              [:td {:title (str col)
-                    :style {:padding-left (str (+ 0.25 level) "rem")
-                            :max-width "16rem"}
-                    :class "pr-1 box-content truncate"}
-               (if (expandable-row? row)
-                 [:button {:class "pr-1 focus:outline-none"
-                           :on-click #(reset! open? (not @open?))}
-                  (if @open? "▼" "▶")]
-                 [:span {:class "pr-1 invisible"} "▶"])
-               (render-col col)]
-              [:td {:title (str col)
-                    :style {:max-width 0
-                            :min-width (when-not full-width? 100)}
-                    :class (if full-width? "pl-3 w-full" "pl-3 truncate w-full")}
-               (render-col col)])]))]
+           (let [row-id {:level level :col col :table-id table-id}]
+             ^{:key (str row-id)}
+             [:<>
+              (if (= 0 i)
+                [:td {:title (str col)
+                      :style {:padding-left (str (+ 0.25 level) "rem")
+                              :max-width "16rem"}
+                      :class "pr-1 box-content truncate"}
+                 (if (expandable-row? row)
+                   (do (when view-state (reset! open? (contains? @view-state row-id)))
+                       [:button {:class "pr-1 focus:outline-none"
+                                 :on-click (fn []
+                                             (when view-state
+                                               (if (contains? @view-state row-id)
+                                                 (swap! view-state disj row-id)
+                                                 (swap! view-state conj row-id)))
+                                             (reset! open? (not @open?)))}
+                        (if @open? "▼" "▶")])
+                   [:span {:class "pr-1 invisible"} "▶"])
+                 (render-col col)]
+                [:td {:title (str col)
+                      :style {:max-width 0
+                              :min-width (when-not full-width? 100)}
+                      :class (if full-width? "pl-3 w-full" "pl-3 truncate w-full")}
+                 (render-col col)])])))]
        (when @open?
          [:tr
           [:td {:col-span (count row) :class "p-0 relative"}
            [tree-table
             (merge props {:level (inc level)
                           :caption nil
-                          :rows (expand-row row)})]
+                          :rows (expand-row row)
+                          :table-id-acc table-id})]
            [:button {:title (str "Collapse  " (pr-str row))
-                     :on-click #(reset! open? false)
+                     :on-click (fn []
```

Comment:
I think we should revert this file and give it one more go from scratch without the extra props. This is getting smelly and I’d rather not take on this tech debt knowing that we’ll have to refactor this in the near future.

Reply:
Agreed